apps/designer/app/routes/$.tsx

```diff
@@ -0,0 +1,128 @@
+import { type LoaderFunction, redirect, MetaFunction } from "@remix-run/node";
+import { useLoaderData } from "@remix-run/react";
+import { InstanceRoot, Root } from "@webstudio-is/react-sdk";
+import { loadCanvasData, type CanvasData } from "~/shared/db";
+import config from "~/config";
+import { db } from "@webstudio-is/project/index.server";
+import env, { Env } from "~/env.server";
+import { sentryException } from "~/shared/sentry";
+import { Canvas } from "~/canvas";
+
+type Mode = "edit" | "preview" | "published";
+
+type Data =
+  | (CanvasData & { env: Env; mode: Mode })
+  | { errors: string; env: Env };
+
+type ProjectIdentifier = { type: "id" | "domain"; value: string };
+
+export const isCanvasRequest = (
+  request: Request
+): false | { projectId: ProjectIdentifier; mode: Mode; pathname: string } => {
+  const url = new URL(request.url);
+
+  const pathname = url.pathname;
+
+  if (Object.values(config).some((path) => pathname.startsWith(path))) {
+    return false;
+  }
+
+  let projectId: ProjectIdentifier | undefined = undefined;
+
+  // @todo all this subdomain logic is very hacky
+  const host =
+    request.headers.get("x-forwarded-host") ||
+    request.headers.get("host") ||
+    "";
+  const [userDomain, wstdDomain] = host.split(".");
+  if (wstdDomain === "wstd" || wstdDomain?.includes("localhost")) {
+    projectId = { type: "domain", value: userDomain };
+  }
+
+  if (projectId === undefined) {
+    const projectIdParam = url.searchParams.get("projectId");
+    if (projectIdParam !== null) {
+      projectId = { type: "id", value: projectIdParam };
```

Comment:
projectId as an object semantically seems a bit weird, because I assumed projectId is the actual id.

Reply:
I would just make the type part of the response object so that there is. {type: 'id', projected, mode, pathname}

Reply:
Renamed `projectId` to `projectIdObject`, hope that's less confusing.

I think `{type: 'id', projected, mode, pathname}` would be also incorrect semantically because `id / domain` is a type of data that is used to identify the project, it's not the type of the entire object that the functions returns, if that makes sense.

Reply:
you could make it 2 separate object depending on if the data is identified by id vs domain ... or maybe there could be some other idea, projectIdObject is clear but not nice

Reply:
I think semantics of this still look weird:
```
    const project =
      projectId.type === "id"
        ? await db.project.loadById(projectId.value)
        : await db.project.loadByDomain(projectId.value);
        ```
        
        Because in one case projectId.value is a domain and in another case it's an id and reading this, you won't clearly see the difference unless you go digging the type

Reply:
This would read much more straight forward
```
    const project =
      'projectId' in parsedUrl
        ? await db.project.loadById(parsedUrl.projectId)
        : await db.project.loadByDomain(parsedUrl.domain);
```

Reply:
Did it as you suggested in the last comment. Good idea!